### PR TITLE
[SMTChecker] Support to Bool variables

### PR DESCRIPTION
**Pull request by leonardoalt**
This PR adds support to Bool program variables including two types of tests: only Bool and mixed Bool/Integer tests.

### CODE REVIEW

File: test/libsolidity/SMTChecker.cpp [hunk 1]
```diff
@@ -329,6 +329,104 @@ BOOST_AUTO_TEST_CASE(ways_to_merge_variables)
...
+					}
+				} else {
+					if (y) {
+						a = 1;
+					} else {
+						a = 0;
+					}
+				}
+				bool xor_x_y = (x && !y) || (!x && y);
+				assert(!xor_x_y || a > 0);
```
Comment (chriseth) [inline test/libsolidity/SMTChecker.cpp:463]: Why not `xor_x_y == (a > 0)`?

File: test/libsolidity/SMTChecker.cpp [hunk 2]
```diff
@@ -329,6 +329,104 @@ BOOST_AUTO_TEST_CASE(ways_to_merge_variables)
...
+
+BOOST_AUTO_TEST_CASE(bool_int_mixed)
+{
+	string text = R"(
+		contract C {
+			function f(bool x) public pure {
+				uint a;
+				if(x)
+					a = 1;
+				assert(!x || a > 0);
```
Comment (chriseth) [inline test/libsolidity/SMTChecker.cpp:430]: Why not `(!x) != (a > 0)`?

File: libsolidity/formal/SMTChecker.cpp [hunk 3]
```diff
@@ -444,7 +444,7 @@ void SMTChecker::arithmeticOperation(BinaryOperation const& _op)
 void SMTChecker::compareOperation(BinaryOperation const& _op)
 {
 	solAssert(_op.annotation().commonType, "");
-	if (_op.annotation().commonType->category() == Type::Category::Integer)
+	if (SSAVariable::supportedType(_op.annotation().commonType.get()))
```
Comment (chriseth) [inline libsolidity/formal/SMTChecker.cpp]: Unrelated to this PR, but shouldn't `supportedType` receive its argument via reference?

**Comment by chriseth** [comment]
Did you test all bool operators?
 - comparison (less than, ...)
 - equality / inequality
 - negation

More or less related: Did we implement short-circuiting correctly?

File: libsolidity/formal/SolverInterface.h [hunk 4]
```diff
@@ -132,10 +133,12 @@ class Expression
 	Expression operator()(Expression _a) const
 	{
 		solAssert(
```
Comment (chriseth) [inline libsolidity/formal/SolverInterface.h:135]: It is probably better to restructure this function to first assert that the arguments are empty, then handle the cases of int and bool function regularly and put a `solAssert(false, ...)` in the final else case.

File: libsolidity/formal/SolverInterface.h [hunk 5]
```diff
@@ -167,9 +170,11 @@ class SolverInterface
 
 	virtual Expression newFunction(std::string _name, Sort _domain, Sort _codomain)
 	{
-		solAssert(_domain == Sort::Int && _codomain == Sort::Int, "Function sort not supported.");
+		solAssert(_domain == Sort::Int && (_codomain == Sort::Int || _codomain == Sort::Bool), "Function sort not supported.");
```
Comment (chriseth) [inline libsolidity/formal/SolverInterface.h]: This could be similarly restructured.

File: test/libsolidity/SMTChecker.cpp (hunk 1 above)
Comment (leonardoalt) [reply test/libsolidity/SMTChecker.cpp:463]: Works too. My initial goal was to have it as an implication: `xor_x_y -> a > 0`, but your example is stronger.

File: test/libsolidity/SMTChecker.cpp (hunk 2 above)
Comment (leonardoalt) [reply test/libsolidity/SMTChecker.cpp:430]: Here the goal was also to have an implication: `x -> a > 0`.

File: libsolidity/formal/SMTChecker.cpp (hunk 3 above)
Comment (leonardoalt) [reply libsolidity/formal/SMTChecker.cpp]: Currently it's via pointer, but I can of course change it to reference if that's preferred.

**Comment by leonardoalt** [comment]
Negation, equality and inequality yes.

What is the behavior for LT, LE, GT and GE? I couldn't find it in the docs.

> Did we implement short-circuiting correctly?

Where exactly? The encoding has the same structure as the original expression, so I would say yes.

File: libsolidity/formal/SolverInterface.h (hunk 4 above)
Comment (leonardoalt) [reply libsolidity/formal/SolverInterface.h:135]: Yes, good idea.

File: libsolidity/formal/SMTChecker.cpp (hunk 3 above)
Comment (chriseth) [reply libsolidity/formal/SMTChecker.cpp]: The general idea in the codebase is to only use a raw pointer if nullptr is a valid argument.

**Comment by chriseth** [comment]
`false < true`.

short circuiting: is `if (x || f())` evaluating `f()` in case `x` is true? It should not, i.e. be roughly equivalent to

```
bool result = false;
if (x) { result = true; }
else if (f()) { result = true; }
```

**Comment by chriseth** [comment]
I don't think we can properly test this at this stage, though.

**Comment by chriseth** [comment]
Perhaps `bool x = false; if (true || (x = true)) {}; assert(!x);` could work.

**Comment by leonardoalt** [comment]
I see your point, but I'm not sure this would make any difference for the solver.
This code
```
bool result = false;
if (x) { result = true; }
else if (f()) { result = true; }
```
says that `x -> result` and `f() -> result`. The solver may still just set both as true.

File: libsolidity/formal/SMTChecker.cpp (hunk 3 above)
Comment (leonardoalt) [reply libsolidity/formal/SMTChecker.cpp]: I'm changing it back to `Type::Category`, not sure why I went to pointers in the first place...

**Comment by leonardoalt** [comment]
- Implemented comparison (gt, ge, lt, le) and tests.
- Replaced `Declaration` pointer by reference in `SSAVariable` and `Symbolic*Variable`.
- Applied the suggested restructure for `SolverInterface`.

File: libsolidity/formal/SMTChecker.cpp [hunk 6]
```diff
@@ -444,21 +445,36 @@ void SMTChecker::arithmeticOperation(BinaryOperation const& _op)
...
+				op == Token::NotEqual ? (left != right) :
+				op == Token::LessThan ? (left < right) :
+				op == Token::LessThanOrEqual ? (left <= right) :
+				op == Token::GreaterThan ? (left > right) :
+				/*op == Token::GreaterThanOrEqual*/ (left >= right)
+			);
+		}
+		else // Bool
+		{
+			value = make_shared<smt::Expression>(
```
Comment (chriseth) [inline libsolidity/formal/SMTChecker.cpp]: Please `solAssert` that it is a bool.

File: libsolidity/formal/SSAVariable.cpp [hunk 7]
```diff
@@ -26,23 +27,35 @@ using namespace dev;
...
 }
 
-bool SSAVariable::supportedType(Type const* _decl)
+bool SSAVariable::supportedType(Type::Category _category)
 {
-	return dynamic_cast<IntegerType const*>(_decl);
+	return typeInteger(_category) || typeBool(_category);
+}
+
+bool SSAVariable::typeInteger(Type::Category _category)
```
Comment (chriseth) [inline libsolidity/formal/SSAVariable.cpp]: `isInteger` / `isBool`?

**Review by chriseth — CHANGES_REQUESTED** [review]
Only tiny changes left. I created a separate issue for the short-circuiting: https://github.com/ethereum/solidity/issues/3711

**Comment by leonardoalt** [comment]
Changes applied.

**Review by chriseth — APPROVED** [review]

**Comment by chriseth** [comment]
There are test failures. Re-running to see if they were infrastructure problems.

**Pull request merged**